uast/uast-encoding-v2.md

```diff
@@ -0,0 +1,249 @@
+# UAST v2 binary encoding
+
+UAST binary encoding uses [Protocol Buffers](https://developers.google.com/protocol-buffers/)
+to encode individual messages, but requires additional logic to reconstruct the UAST.
+
+The design document can be found [here](https://docs.google.com/document/d/1iDOuWLnTckBdSscUMWGl2hBqAAqRn8jdRabXZSdr2aw/edit?usp=sharing)
+and the reference implementation in Go can be found [in SDK](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodesproto.go).
+
+On a high level, a binary UAST uses [UAST v2 data model](./representation-v2.md).
+It extends this model by adding an ephemeral node ID field used to link UAST nodes.
+
+The file consist of two sections:
+
+* [Header](#header), containing "magic", format version and other options.
+* Flat [list of nodes](#nodes) linked by node IDs.
+
+The list contains zero or more graph nodes addressed by ID. Binary format version `1` must
+only store tree structures, but the format was made compatible with graph structure for future
+extensions.
+
+The ID of the root tree node is stored in the [Header](#header) message. Optionally an ID
+of a second tree root can be provided to store file metadata. Implementations are free to ignore it.
+
+## Encoding
+
+### Messages
+
+Each protocol buffer message in the binary format is length-delimited, similar to `bytes` values used in protocol buffers fields.
+
+First, a message size is encoded as an [unsigned varint](https://developers.google.com/protocol-buffers/docs/encoding),
+followed by `N` bytes of the message itself.
+
+There are only two messages used to encode the UAST:
+
+- [GraphHeader](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L17)
+- [Node](https://github.com/bblfsh/sdk/blob/b44cc02214bf0d9c55dc3b830f423a6df6709ca9/uast/nodes/nodesproto/nodes.proto#L46)
+
+See the sections below for the usage of those messages.
+
+### Header
+
+The binary UAST file begins with a `\x00bgr` (`0x00 0x62 0x67 0x72`) "magic" that can be
+used to detect a format of the file.
+
+It is followed by a version number encoded as an unsigned 32 bit integer in little-endian byte order.
+The current format version is `1`.
+
+Next, a length-delimited `GraphHeader` message follows:
+```proto
+message GraphHeader {
+    uint64 last_id  = 1;
+    uint64 root     = 2;
+    uint64 metadata = 3;
+}
+```
+
+`last_id` is an optional field that sets a maximal node ID reserved by the node ID allocator.
+If not set, `max(node.ID)` is assumed. Implementation may reserve more IDs than required by
+the UAST nodes by setting `last_id > max(node.ID)`.
+
+`root` is an ID of the tree root node. Multiple unnamed roots can be stored by referencing an
```

Comment:
Sure, we use Arrays in the AST in general. But if you mean the array-of-roots, then no, we always specify a single root node in Babelfish.

This was done for a few reasons. First, we wanted a future-proof format (graphs = multiple roots =  no explicit root). But at the same time, we had a requirement from management that simple implementation of the decoder should be allowed to support trees only.

Thus, the format itself allows both behaviors: Babelfish emits the root pointer for compatibility, but other tools might not. Because of this, the spec is a bit vague here. It expects super naive implementation to fail without an explicit root, but more advanced implementations (that still expect a tree) can instead calculate an "array of roots" and use it as an implicit root node.

I agree that having this ambiguity is not ideal and we can discuss it separately.

As a note, if we decide to drop it, the Babelfish decode was always prepared for this and even old versions will automatically detect the root.

Reply:
Re: Arrays in general vs. in the root. Yes, sorry for the ambiguity, I meant "do we use it in this context?"

It seems to me like we can simplify the spec, retain future-proofing, and reduce ambiguity for the implementor, by mandating that the reader must always scan for roots. For example maybe:

1. Any node in the list that is not referenced by any (other) node in the list is considered a _root candidate_.
2. If the `root` field of the header gives the ID of a root candidate, that candidate is treated as the root and the other root candidates are ignored.
2. If the `root` field of the header is 0, [1]
3. Otherwise (the `root` field does not point to a root candidate) the reader must report an error.

[Note 1]: There are options here: My intuition would be to say this is an error for now, and later when we add more graph structure we can relax that. That way a reader that expects one tree will fail predictably when we add a forest of trees and/or more graph structure, but can grow to accommodate it cleanly. 

Another option is to say that if `Header.root == 0` and there is exactly 1 root candidate, that must be used as the root, and if there are 0 root candidates or > 1 root candidates, it is an error. This is probably a good compatibility choice.

In any case, what I'm hoping to get at here is a clear specification of what the reader can expect. Right now we kind of allow "anything goes", there might be multiple roots or might not, and there might be a pointer to one of them, or to an array of them, and there's a bunch of correctness checks in there (do all the elements of the array point to root candidates? But then they're referenced inside the nodes list and aren't really root candidates; except that maybe this array is special because it's pointed to by `Header.root`? And so on).

Reply:
I like the way you described the scan for root, but I have a few notes:
* Assuming future compatibility with graphs, the `root` may not be in the list of root candidates (it may be referenced by another graph node).
* If `root` is zero (`nil` tree), there is still a case when `metadata` may not be zero. If we won't mention `metadata`, the implementation will incorrectly pick it as root (since it's the only root candidate).